include/picongpu/plugins/binning/Binner.hpp

```diff
@@ -0,0 +1,319 @@
+                const T_Worker& worker,
+                const T_DataSpace& extentsDataspace,
+                const T_AxisKernelTuple& axisKernelTuple,
+                T_DataBox histBox) const
+            {
+                // TODO check normDataBox shape is same as histBox
+                auto blockIdx = cupla::blockIdx(worker.getAcc()).x * blockSize;
+                auto forEachElemInDataboxChunk = lockstep::makeForEach<blockSize>(worker);
+                forEachElemInDataboxChunk(
+                    [&](int32_t const linearIdx)
+                    {
+                        int32_t const linearTid = blockIdx + linearIdx;
+                        if(linearTid < extentsDataspace.productOfComponents())
+                        {
+                            pmacc::DataSpace<nAxes> const nDIdx
+                                = pmacc::DataSpaceOperations<nAxes>::map(extentsDataspace, linearTid);
+                            float_X factor = 1.;
+                            apply(
+                                [&](auto const&... tupleArgs)
+                                {
+                                    // uses bin width for axes without dimensions as well should those be ignored?
+                                    uint32_t i = 0;
+                                    ((factor *= tupleArgs.getBinWidth(nDIdx[i++])), ...);
+                                },
+                                axisKernelTuple);
+
+                            histBox(linearTid) *= 1 / factor;
+                        }
+                    });
+            }
+        };
+
+        void dimensionSubtraction(std::array<double, 7>& outputDims, const std::array<double, 7>& axisDims)
+        {
+            for(size_t i = 0; i < 7; i++)
+            {
+                outputDims[i] += -axisDims[i];
+            }
+        }
+
+        template<typename TBinningData>
+        class Binner : public INotify
+        {
+            using TDepositedQuantity = typename TBinningData::DepositedQuantityType;
+
+        private:
+            TBinningData binningData;
+            MappingDesc* cellDescription;
+            std::unique_ptr<HostDeviceBuffer<TDepositedQuantity, 1>> histBuffer;
+            uint32_t accumulateCounter = 0;
+            mpi::MPIReduce reduce{};
+            bool isMaster = false;
+            WriteHist histWriter;
+
+        public:
+            Binner(TBinningData bd, MappingDesc* cellDesc) : binningData{bd}, cellDescription{cellDesc}
+            {
+                Environment<>::get().PluginConnector().setNotificationPeriod(this, binningData.notifyPeriod);
+
+                /**
+                 * Allocate and manage global histogram memory here, to facilitate time averaging
+                 */
+
+                this->histBuffer = std::make_unique<HostDeviceBuffer<TDepositedQuantity, 1>>(
+                    binningData.axisExtentsND.productOfComponents()); // TODO for auto n_bins. allocate full size
```

Comment:
please move the comment into the comment block above.

Reply:
Moved